Ant Colony Optimization

zjost opened the pull request:
See the README for more details.  Issue: #150.

The screenshot below shows the interactive visualization with the two main parameters (alpha and beta) as float sliders.  We show the initial network grid (a fully connected set of cities), the best solution length found per iteration, the best global solution length found, and the city layout with the pheromone trail strength indicated by edge width.

<img width="1691" alt="image" src="https://github.com/user-attachments/assets/8d828fb2-d97c-4920-86d5-d306bf30863c">

EwoutH commented:
Thanks for opening a PR to share your model! This looks like a very high effort PR, I will look tomorrow morning (CET) into it in depth.

In the mean time, a bit more extensive Readme and PR start would be useful (possibly with some screenshots), and I’ve enabled the CI to run to detect any potential errors.

zjost commented:
I'm seeing a unit test failure:
```
TypeError: AcoTspModel.__init__() missing 2 required positional arguments: 'num_agents' and 'tsp_graph'
```

I see a build error because the test assumes the model can be constructed with no passed parameters.  I went ahead and updated that, so a new run should pass that test.

wang-boyu commented:
Thanks for your efforts. Do you mind updating the top-level README file, to include your example under section `Network Examples`?

EwoutH commented on examples/aco_tsp/run_tsp.py:
```diff
@@ -0,0 +1,47 @@
+
+    _, ax = plt.subplots()
+    ax.plot(results["best_distance"])
+    ax.set(xlabel="Episode", ylabel="Best distance", title="Best distance per episode")
```
Somehow this doesn't work, the title and labels don't show up. Maybe some of the [other examples](https://github.com/search?q=repo%3Aprojectmesa%2Fmesa-examples%20title&type=code) can help (I see everyone does it differently).

rht replied:
No, the cause is because the axis is initialized using the global non-thread-safe method, `plt.*`. It has to be done this way using the OOP API: https://github.com/projectmesa/mesa/blob/c96ee458ca480c786802587fa4e8d0d997b4de95/mesa/visualization/components/matplotlib.py#L11-L12.

rht replied:
Oh, is this not the SolaraViz? Then my comment above is relevant.

EwoutH replied:
Sorry, it was in SolaraViz, but I placed the review comment on the wrong place.

zjost replied:
When using SolaraViz, I use the OOP API.  This comment is for the script that doesn't use Solara.  Either way, I see the axis labels as intended.  Which plots are unlabelled for you?

EwoutH commented on examples/aco_tsp/setup.py:
Does this imply that you example is an actual package? Not that I'm against it, just curious.

There's only one [other example](https://github.com/projectmesa/mesa-examples/blob/main/gis/agents_and_networks/setup.py) that does this, @rht and @wang-boyu, what's the best practice here / what do we want for Mesa-examples?

wang-boyu replied:
I don't really have any preferences about whether a model should be packaged or not. The Agents-and-Networks model was packaged mainly because there were too many classes and functions. There used be another notebook folder containing jupyter notebooks, and I found it easier to import functions into notebooks if the model was packaged.

zjost replied:
This file was auto-generated by running `mesa startproject` as defined in Mesa's [Best Practices](https://mesa.readthedocs.io/en/latest/best-practices.html).  I'm happy to delete it though, because I'm not using it.  I'll leave it up to you guys.

EwoutH commented on examples/aco_tsp/run_tsp.py:
```diff
@@ -0,0 +1,47 @@
+        results["best_distance"].append(model.best_distance)
+        results["best_path"].append(model.best_path)
+        print(
+            f"Episode={e + 1}; Min. distance={model.best_distance}; pheromone_1_8={model.grid.G[17][15]['pheromone']}"
```
Prints nicer with a few less decimals
```suggestion
            f"Episode={e + 1}; Min. distance={model.best_distance:.2f}; pheromone_1_8={model.grid.G[17][15]['pheromone']:.4f}"
```

EwoutH commented on examples/aco_tsp/run_tsp.py:
```diff
@@ -0,0 +1,47 @@
+        if model.best_distance < best_distance:
+            best_distance = model.best_distance
+            best_path = model.best_path
+            print(f"New best distance:  distance={best_distance}")
```
```suggestion
            print(f"New best distance: {best_distance:.2f}")
```

EwoutH commented on examples/aco_tsp/app.py:
I really like the visualisation! The only thing I'm curious about if we can make the Network displayed somewhat more useful, by color coding the nodes and/or edges or something.

zjost replied:
I thought about that.  The only way I could think of to make it useful was to force the nodes into their geometric positions (since they represent cities on a map), and have edge widths proportional to something meaningful.  That is what led to the custom visualization, because I had to hack on the mesa source code to be able to assign node positions.  [This Mesa code](https://github.com/projectmesa/mesa/blob/c96ee458ca480c786802587fa4e8d0d997b4de95/mesa/visualization/components/matplotlib.py#L65) forces you to use a spring layout.  I think there are three options:
- Get rid of the plot (I'm not sure how to do that)
- Think of some other visualization
- Figure out a way to do my custom plot, but in Mesa.  This might require a Mesa source code change.

EwoutH reviewed (commented):
Really awesome work! I have a few minor points on the visualisation, prints and packaging.

Reviewing the model code next.

EwoutH commented on examples/aco_tsp/README.md:
Please remove the newlines within paragraphs, so a paragraph is just a single line, markdown files wrap them automatically.

EwoutH commented on examples/aco_tsp/README.md:
```diff
@@ -0,0 +1,53 @@
+In other words, $\alpha$ and $\beta$ are tuned to set the relative importance
+of the phermone trail left by prior ants, and the greedy heuristic of
+1-over-distance.
+
```
Maybe add two section:
- What's notable or remarkable about this model from a Mesa perspective (I think at least the use of a `NetworkGrid`)?
- What's can uses play/experiment with? (input variables, etc.)

EwoutH commented on examples/aco_tsp/aco_tsp/model.py:
```diff
@@ -0,0 +1,249 @@
+            results.append(val)
+
+        results = np.array(results)
+        Z = results.sum()
```
It's a bit uncommon to use a capital letter for a variable (it's usually reserved for classes).

EwoutH commented on examples/aco_tsp/aco_tsp/model.py:
```diff
@@ -0,0 +1,249 @@
+        self.best_distance = float("inf")
+        self.best_distance_iter = float("inf")
+
+        self.datacollector = mesa.datacollection.DataCollector(
```
Maybe describe in the Readme briefly what data is collected and how that is used or can be used to gain insights into the system.

EwoutH reviewed (commented):
Code quality looks really good, and is very suitable as an example model. I have a few minor nitpicks, but nothing major.

To strengthen the example, I think the Readme could be extended to point to the highlights of this example.

zjost commented:
Thank you all for the thorough review!  I'm about to leave for a week on a vacation, but will implement all of these changes when I return.

zjost commented:
I think I've addressed all the comments, except there are perhaps 2 open questions with the visualizations:
- Is there something we can do to [make the "Space" plot more useful](https://github.com/projectmesa/mesa-examples/pull/157#discussion_r1715123677)?
- Is there something wrong with the [plotting axis titles](https://github.com/projectmesa/mesa-examples/pull/157#discussion_r1715111352)?

From my perspective, ready to merge when you all are satisfied.  Thanks again!

EwoutH commented:
Thanks a lot, it's looking really good!

> * Is there something wrong with the [plotting axis titles](https://github.com/projectmesa/mesa-examples/pull/157#discussion_r1715111352)?

I meant in the Solara visualzation, it's not showing up for me:

![image](https://github.com/user-attachments/assets/faf39af3-2d60-4886-813a-75afec433423)

EwoutH commented:
> - Is there something we can do to [make the "Space" plot more useful](https://github.com/projectmesa/mesa-examples/pull/157#discussion_r1715123677)?

Maybe we can color the cities on the percentage of agents visited? For example with a `'viridis'` [colormap](https://matplotlib.org/stable/users/explain/colors/colormaps.html)? That might give a very nice visualisation.

EwoutH commented:
To be clear by the way: This example is already in excellent shape and I'm ready to merge it at any point. Right now I'm just trying to see if we can get the the icing on the cake, because now this example is fresh in our minds.

If you're happy with it the way it is, just say so, and I will merge it and make the final improvements myself.

zjost commented:
> I meant in the Solara visualzation, it's not showing up for me:

I see what you mean now, and I observe the same thing.  For the line plots, I just specify them in the `measures` field as class attributes, i.e.:
```python
page = SolaraViz(
    AcoTspModel,
    ...
    measures=["best_distance_iter", "best_distance", ...],
    ...
)
```
Is there a standard way to set x-axis and titles in that case?  I would expect Mesa to set a default through this API.

> > * Is there something we can do to [make the "Space" plot more useful](https://github.com/projectmesa/mesa-examples/pull/157#discussion_r1715123677)?
> 
> Maybe we can color the cities on the percentage of agents visited? For example with a `'viridis'` [colormap](https://matplotlib.org/stable/users/explain/colors/colormaps.html)? That might give a very nice visualisation.

In this case, all ants visit every city every time, so I'm not sure that would be instructive.  I'll take your suggestion to just remove the space viz--thanks!

Regarding when to merge, I'm happy to address any outstanding little issues and nits.  As you say, it has our attention now, so now's the time to get it right.  

Since I'm removing the space plot, it seems we just get to get the remaining plot axes/titles sorted.  I added a title to the custom viz, but am not sure what to do on the measures.

EwoutH commented:
> Is there a standard way to set x-axis and titles in that case? I would expect Mesa to set a default through this API.

Yeah I also noticed that, and I don't think there is (@rht could you maybe confirm)?

I will open a new issue for that, and obviously out of scope for this PR.

> Since I'm removing the space plot, it seems we just get to get the remaining plot axes/titles sorted. I added a title to the custom viz, but am not sure what to do on the measures.

That's perfect!

rht commented:
> Yeah I also noticed that, and I don't think there is (@rht could you maybe confirm)?

Only the y axis is labeled. https://github.com/projectmesa/mesa/blob/3cf1b76ca3a157c8a064b60130755a86e73718be/mesa/visualization/components/matplotlib.py#L123. IIRC it was to mimic the old Tornado viz. We could add the x label ("steps"), but the title would be redundant with the y label.

zjost commented:
Is this ready to merge, or are we still waiting on something?

rht commented:
Looks fine to me.

EwoutH commented:
@zjost Thanks, awesome to have this new example in the repo!
